Sources/MongoSwift/Logging.swift

```diff
@@ -0,0 +1,132 @@
+            fatalError("Unrecognized libmongoc log level \(self)")
+        }
+    }
+}
+
+/// If the env var has been set to enable logging, creates a corresponding Logger with the specified level. Else,
+/// returns nil.
+private func makeLogger(envVar: String, label: String) -> Logger? {
+    // No level was provided.
+    guard let envVarLevel = ProcessInfo.processInfo.environment[envVar]?.lowercased() else {
+        return nil
+    }
+    // "off" or any unrecognized level. (should we do something different on an unrecognized level?)
+    guard let level = Logger.Level(rawValue: envVarLevel) else {
+        return nil
+    }
+
+    var logger = Logger(label: label)
+    logger.logLevel = level
+    return logger
+}
+
+/// Global loggers corresponding to each component. We will likely add more of these with time as more components are
+/// defined in the logging spec.
+private let COMMAND_LOGGER = makeLogger(envVar: "MONGODB_LOGGING_COMMAND", label: "MongoSwift.command")
+private let SDAM_LOGGER = makeLogger(envVar: "MONGODB_LOGGING_SDAM", label: "MongoSwift.sdam")
+private let SERVER_SELECTION_LOGGER = makeLogger(
+    envVar: "MONGODB_LOGGING_SERVER_SELECTION",
+    label: "MongoSwift.serverSelection"
+)
+private let CONNECTION_LOGGER = makeLogger(envVar: "MONGODB_LOGGING_CONNECTION", label: "MongoSwift.connection")
+
+/// For testing purposes, we define fallback loggers which can be mutated as needed to temporarily turn logging on/off
+/// during tests. Arguably you could just handle this by making the original variables mutable, but it seemed a bit
+/// safer to keep those locked down with their initial values.
+internal var TEST_COMMAND_LOGGER: Logger?
+internal var TEST_SDAM_LOGGER: Logger?
+internal var TEST_SERVER_SELECTION_LOGGER: Logger?
+internal var TEST_CONNECTION_LOGGER: Logger?
+
+extension mongoc_structured_log_component_t {
+    /// Returns the Swift logger corresponding to the mongoc log component. This value is nil if the user did not
+    /// enable logging for the component, or if the component is unrecognized.
+    fileprivate var swiftLogger: Logger? {
+        switch self {
+        case MONGOC_STRUCTURED_LOG_COMPONENT_COMMAND:
+            return COMMAND_LOGGER ?? TEST_COMMAND_LOGGER
+        case MONGOC_STRUCTURED_LOG_COMPONENT_SDAM:
+            return SDAM_LOGGER ?? TEST_SDAM_LOGGER
+        case MONGOC_STRUCTURED_LOG_COMPONENT_SERVER_SELECTION:
+            return SERVER_SELECTION_LOGGER ?? TEST_SERVER_SELECTION_LOGGER
+        case MONGOC_STRUCTURED_LOG_COMPONENT_CONNECTION:
+            return CONNECTION_LOGGER ?? TEST_CONNECTION_LOGGER
+        default:
+            // ignore an unrecognized component. (TODO: codify in spec what should happen here?)
+            return nil
+        }
+    }
+}
+
+/// Callback for handling a mongoc structured log message. This is registered one time via `initializeMongoc()`.
+internal func handleMongocStructuredLogMessage(entry: OpaquePointer!, _: UnsafeMutableRawPointer?) {
```

Comment:
what is the second argument that we ignore here? could we name it something to describe it (still with underscore prefix though)?

Reply:
So the second parameter to `mongoc_structured_log_set_handler` is an `UnsafeMutableRawPointer`, which is basically a pointer to anything. the idea is that you could use it to register some contextual object, which is then passed into the callback here whenever it's executed.

This is similar to the parameter accepted by `mongoc_client_pool_set_apm_callbacks` [here](https://github.com/mongodb/mongo-swift-driver/blob/c3125ffee1919fb758cd7c8d3f04f77cb804eeda/Sources/MongoSwift/ConnectionPool.swift#L237), where we pass in the `MongoClient` so that when executing the callback we know which client it was associated with.

In this case since the callback is global rather than per-client, all context can just be tracked globally in Swift and I didn't find a need to use it.

swiftformat seems to want to use `_` for the parameter name since it's unused, so I've added `context` as a label and kept `_` as the parameter name.